modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityStressUpdateBase.C

```diff
@@ -329,25 +329,37 @@ CrystalPlasticityStressUpdateBase::calculateSchmidTensor(
 }
 
 void
-CrystalPlasticityStressUpdateBase::calculateShearStress(RankTwoTensor & pk2)
+CrystalPlasticityStressUpdateBase::calculateShearStress(
+    const RankTwoTensor & pk2, const RankTwoTensor & inverse_thermal_deformation_grad)
 {
+  RankTwoTensor thermal_deformation_grad = inverse_thermal_deformation_grad.inverse();
   for (unsigned int i = 0; i < _number_slip_systems; ++i)
-    _tau[_qp][i] = pk2.doubleContraction(_flow_direction[_qp][i]);
+  {
+    // compute PK2_hat using thermal deformation gradient
+    RankTwoTensor pk2_hat = thermal_deformation_grad.det() * thermal_deformation_grad.transpose() *
+                            pk2 * inverse_thermal_deformation_grad.transpose();
+    _tau[_qp][i] = pk2_hat.doubleContraction(_flow_direction[_qp][i]);
+  }
 }
 
 void
 CrystalPlasticityStressUpdateBase::calculateTotalPlasticDeformationGradientDerivative(
-    RankFourTensor & dfpinvdpk2, const RankTwoTensor & inverse_plastic_deformation_grad_old)
+    RankFourTensor & dfpinvdpk2,
+    const RankTwoTensor & inverse_plastic_deformation_grad_old,
+    const RankTwoTensor & inverse_thermal_deformation_grad_old)
```

Comment:
Those codes seem to be hard to extend to other eigenstrains, e.g., irradiation eigenstrain. Do you have any ideas how to include genetic eigenstrains? This is not required for this PR, but it will be nice to have some thought here.

Reply:
Glad that you asked! I think possible to make it more flexible in terms of additional eigenstrain types. I have to re-derive some equations further in terms of how this affects the Jacobian calculations. But from what I can see so far, it is possible to have one deformation gradient, e.g., F_eigen = F_th * F_irradiation * ..., in the base class, which encompasses all the additional eigenstrain contributions.  

I will definitely need to discuss with you more once I start extending this capability.